File: project5/ruby_work/linked_list.rb
# Tyler Hansen, Sawyer Strong, Skye Rhomberg
# CS333 Spring 2021 -- Programming Languages
# 03/31/2021

# linked_list.rb: Describes the singly-linked list data structure.


class LinkedList

    # Get and Set Instance variables head, tail, and length
    attr_accessor :head, :tail, :length

    # Function: initialize
    # ----------------------
    # The constructor for a LinkedList object.
    # 
    # Parameters: N/A
    #
    # Returns: N/A
	def initialize
		self.head   = nil
		self.tail   = nil
		self.length = 0
	end

	# Function: append
    # ----------------------
    # Append a new node to the bottom of the list.
    # 
    # Parameters:
    # data -- The data stored in the new node.
    #
    # Returns: N/A
	def append data
        # Create a new Node
		node = Node.new data
		unless head # Unless there is a head node, set the head to the new node.
			self.head = node
        end
		# Set the next of the tail to the new node. 
		current_tail = self.tail
        self.tail.next = node
		self.tail = node
		self.length += 1
	end

	# Function: delete
    # ----------------------
    # Delete a node from the list.
    # 
    # Parameters:
    # node -- The node to remove (actually a node object, not the data stored at that node).
    #
    # Returns: N/A
	def delete node # Takes in a node

		return nil unless node # Return nil if no node is passed in

        # If it is the head node...
		if node == head
			if head.next.nil?
				self.head = self.tail = nil
			else
				self.head = self.head.next
			end
		else # If it is not the head node...
			tmp = self.head
			while tmp && tmp.next != node
				tmp = tmp.next
			end
			tmp.next = node.next if tmp
		end
		self.length -= 1
	end

	# Function: clear
    # ----------------------
    # Remove all nodes from the list.
    # 
    # Parameters: N/A
    #
    # Returns: N/A
	def clear
		while self.head.next != nil
            remove self.head.next.data
        end
        remove self.head.data
        self.length = 0
	end


    # Function: remove
    # ----------------------
    # The constructor for a Node object.
    # 
    # Parameters:
    # value -- The data in the node to remove.
    #
    # Returns: NA
    def remove(value)
        if self.head.data == value
          self.head = self.head.next
          return
        end
        node = find(value)
        node.next = node.next.next
        self.length -= 1
    end

    # Function: find
    # ----------------------
    # Get a node by value.
    # 
    # Parameters:
    # value -- The data in the node to seek.
    #
    # Returns:
    # node -- The node with the found value or false
    def find(value)
        node = self.head
        return false if !node.next
        return node  if node.next.data == value
        while (node = node.next)
          return node if node.next && node.next.data == value
        end
    end

    # Function: is_empty?
    # ----------------------
    # Returns whether or not the LinkedList is empty.
    # 
    # Parameters: N/A
    #
    # Returns: true if empty, false otherwise
    def is_empty?
        if self.head == nil
            return true
        else
            return false
        end
    end

    # Function: push
    # ----------------------
    # Push a node on top of the list.
    # 
    # Parameters:
    # data -- the data in to store in the new head node.
    #
    # Returns: N/A
    def push(data)
        new_node = Node.new(data)
        if self.is_empty?
            self.tail = new_node
        else
            current_head = self.head
            new_node.next = current_head
        end
        self.head = new_node
        self.length += 1
    end

    # Function: pop
    # ----------------------
    # Remove and return the top node in the list.
    # 
    # Parameters: N/A
    #
    # Returns: N/A
    def pop
        if self.is_empty?
            return "This list is currently empty"
        else
            current_node = self.head
            while current_node.next.next != nil
                current_node = current_node.next
            end
            last_node = current_node.next
            current_node.next = nil
        end
        return last_node
        self.length -= 1
    end

    # Function: pretty_print
    # ----------------------
    # Print all of the elements in the list as a nice, pretty array.
    # 
    # Parameters: N/A
    #
    # Returns: N/A
    def pretty_print
        puts "Head"
        array = []
        current_node = self.head
        if self.is_empty?
            puts "Tail"
            return array
        else
            while current_node.next != nil
                array << current_node.data
                current_node = current_node.next
            end
            array << current_node.data
       end
       array.append(self.tail.data)
       puts array
       puts "Tail"
    end

    class Node
        # Get and Set instance variables next and data
        attr_accessor :next, :data
    
        # Function: initialize
        # ----------------------
        # The constructor for a Node object.
        # 
        # Parameters:
        # data -- The data to store in the node.
        #
        # Returns: NA
        def initialize data
            self.data = data
            self.next = nil
        end
    end

end

File: project5/c_work/linkedlist.c
/* 
Tyler Hansen
CS333 Spring 2021 -- Programming Languages
3/30/2021
*/

/* 
linked_list.c: Describes a linked list data structure.
*/

#include "linkedlist.h"
#include <stdlib.h>
#include <stdio.h>

/*
 * Function:  ll_create 
 * --------------------
 * Creates a Linked List in memory
 *
 * returns: the new LinkedList object.
 */
LinkedList *ll_create(){
    LinkedList *list = malloc(sizeof(LinkedList)); 
    list->size = 0;
    list->head = NULL;
    list->tail = NULL;
    return list;
}

/*
 * Function:  ll_push 
 * --------------------
 * Adds a node to the front of the list, storing the given data in the node.
 * 
 * LinkedList *l: the LinkedList to add a node to the front of. 
 * void *data: data stored in the node to add to the LinkedList.
 *
 * returns: N/A
 */
void ll_push(LinkedList *l, void *data){
    Node* new_node = malloc(sizeof(Node)); // Allocate space for a new node
    new_node->val = data; // Set new node's data to input param
    // If list is empty, assign tail
    if(l->tail == NULL){
        l->tail = new_node;
    }
    // If list is not empty, make backwards link
    if(l->head != NULL){
        l->head->prev = new_node;
    }
    // Reassignment of head
    new_node->next = l->head;
    l->head = new_node;
    l->size += 1;
    return;
}

/*
 * Function:  ll_pop 
 * --------------------
 * Removes the node at the front of the list and returns the associated data.
 * 
 * LinkedList *l: the LinkedList to remove the first value of. 
 *
 * returns: 
 * data: the data stored in the first node.
 */
void *ll_pop(LinkedList *l){
    if (l->head == NULL){
        printf("Tried to pop empty list...\n");
        exit(-1);
    }
    Node *prev_head = l->head; // Save the current head
    void *return_data = prev_head->val;
    l->head = l->head->next; // Set the new head to the next node in the LinkedList.
    l->head->prev = NULL; // Set the new head's previous to null as it wouldn't be the head if it had a previous.
    l->size -= 1;
    free(prev_head);
    if(l->head == NULL){
        l->tail = NULL;
    }
    return return_data;
} 

/*
 * Function:  ll_append 
 * --------------------
 * Adds a node to the end of the list, storing the given data in the node.
 * 
 * LinkedList *l: the LinkedList to add a node to the front of. 
 * void *data: data stored in the node to add to the LinkedList.
 *
 * returns: N/A
 */
void ll_append(LinkedList *l, void *data){
    Node* new_node = malloc(sizeof(Node));
    new_node->val = data;
    new_node->next = NULL;
    // If list is empty, assign head
    if(l->head == NULL){
        l->head = new_node;
    } 
    // If list is not empty, make forwards link
    if(l->tail != NULL){
        l->tail->next = new_node;
    }
    // Reassignment of tail
    new_node->prev = l->tail;
    l->tail = new_node;
    l->size += 1;
    return;
}

/*
 * Function:  ll_remove
 * --------------------
 * Removes the first node in the list whose data matches target given the comparison function. 
 * The function returns the pointer to the data.
 * 
 * LinkedList *l: the LinkedList we wish to remove an element from.
 * void *target: the target data. 
 * int *compfunc: the function that compares the target data to the data stored in the linked list and returns 1 if they are equal and 0 otherwise.
 *
 * returns: The pointer to the data if found, NULL otherwise.
 */
void *ll_remove(LinkedList *l, void *target, int (*compfunc)(void *, void *)){
    Node* n = l->head;
    // Null return for empty lists
    if (n == NULL){
        return n;
    }
    // Iterate through each node and check for match
    while (n != NULL && !compfunc(n->val, target)) {
        n = n->next;
    }
    // If not found in list, return null
    if (n == NULL){
        return NULL;
    }
    if (n->prev != NULL){
        // New head is not necessary
        n->prev->next = n->next;
    } else {
        // Reset head
        l->head = n->next;
    }
    if(n->next != NULL){
        // New tail is not necessary
        n->next->prev = n->prev;
    } else{
        // Reset tail
        l->tail = n->prev;
    }
    void * return_data = n->val;
    free(n);
    l->size -= 1;
    return return_data;
}

/*
 * Function:  ll_size
 * --------------------
 * Returns the size of the LinkedList
 * 
 * LinkedList *l: the LinkedList we wish to get the size of.
 *
 * returns: 
 * int size: the number of elements in the LinkedList. 
 */
int ll_size(LinkedList *l){
    return l->size;
}

/*
 * Function:  ll_clear
 * --------------------
 * Removes all of the nodes from the list, freeing the associated data using the given function.
 * 
 * LinkedList *l: the LinkedList we wish to clear.
 * *freefunc: the function to free the memory.
 *
 * returns:  N/A
 */
void ll_clear(LinkedList *l, void (*freefunc)(void *)){
    Node* n;
    while (l->head != NULL) {
        n = l->head;
        l->head = l->head->next;
        freefunc(n->val);
        free(n);
    }
    l->head = NULL;
    l->tail = NULL;
    l->size = 0;
    return;
}

/*
 * Function:  ll_map
 * --------------------
 * Traverses the list and applies the given function to the data at each node.
 * 
 * LinkedList *l: the LinkedList we wish to clear.
 * *mapfunc: the function to apply to the data at each node.
 *
 * returns:  N/A
 */
void ll_map(LinkedList *l, void (*mapfunc)(void *)){
    Node* n = l->head; 
    while (n != NULL) {
       mapfunc(n->val);
       n = n->next;
    }
    return;
}